kernel/src/features/mod.rs

```diff
@@ -0,0 +1,233 @@
+    V2Checkpoint,
+    /// Iceberg compatibility support
+    IcebergCompatV1,
+    /// Iceberg compatibility support
+    IcebergCompatV2,
+}
+
+impl TryFrom<String> for WriterFeatures {
+    type Error = Error;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        value.as_str().try_into()
+    }
+}
+
+impl std::str::FromStr for WriterFeatures {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        s.try_into()
+    }
+}
+
+impl TryFrom<&str> for WriterFeatures {
+    type Error = Error;
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        let val = match value {
+            "appendOnly" => WriterFeatures::AppendOnly,
+            "invariants" => WriterFeatures::Invariants,
+            "checkConstraints" => WriterFeatures::CheckConstraints,
+            "changeDataFeed" => WriterFeatures::ChangeDataFeed,
+            "generatedColumns" => WriterFeatures::GeneratedColumns,
+            "columnMapping" => WriterFeatures::ColumnMapping,
+            "identityColumns" => WriterFeatures::IdentityColumns,
+            "deletionVectors" => WriterFeatures::DeletionVectors,
+            "rowTracking" => WriterFeatures::RowTracking,
+            "timestampNtz" => WriterFeatures::TimestampWithoutTimezone,
+            "domainMetadata" => WriterFeatures::DomainMetadata,
+            "v2Checkpoint" => WriterFeatures::V2Checkpoint,
+            "icebergCompatV1" => WriterFeatures::IcebergCompatV1,
+            "icebergCompatV2" => WriterFeatures::IcebergCompatV2,
+            _ => return Err(Error::generic(format!("Unknown writer feature: {}", value))),
+        };
+        Ok(val)
+    }
+}
+
+impl AsRef<str> for WriterFeatures {
+    fn as_ref(&self) -> &str {
+        match self {
+            WriterFeatures::AppendOnly => "appendOnly",
+            WriterFeatures::Invariants => "invariants",
+            WriterFeatures::CheckConstraints => "checkConstraints",
+            WriterFeatures::ChangeDataFeed => "changeDataFeed",
+            WriterFeatures::GeneratedColumns => "generatedColumns",
+            WriterFeatures::ColumnMapping => "columnMapping",
+            WriterFeatures::IdentityColumns => "identityColumns",
+            WriterFeatures::DeletionVectors => "deletionVectors",
+            WriterFeatures::RowTracking => "rowTracking",
+            WriterFeatures::TimestampWithoutTimezone => "timestampNtz",
+            WriterFeatures::DomainMetadata => "domainMetadata",
+            WriterFeatures::V2Checkpoint => "v2Checkpoint",
+            WriterFeatures::IcebergCompatV1 => "icebergCompatV1",
+            WriterFeatures::IcebergCompatV2 => "icebergCompatV2",
+        }
+    }
+}
+
+impl std::fmt::Display for WriterFeatures {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.as_ref())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_roundtrip_reader_features() {
```

Comment:
Round trip tests are good, but they can only validate the cases we remember to add test cases for... is there a way to future-proof them? (see other comment about maybe deriving both directions, so they can't get out of sync)

Reply:
yeah - so we had similar discussions when we created to config system for object_store. Essentially the thing we wanted to at least centralize the logic, so that we could only break in one place. unfortunately we did not fond a satisfying way to safeguard against typos et. al.

Initially I thought we would need to have multiple strings on the parsing side, which is why I did not use strum. however here that does not apply. WE will need to consider this when we are parsing the configuration though. IIRC for append only both `appendOnly` and `delta.appendOnly` are permissible keys.

So here I plan on adding strum as well, but we'll have the same question again when we do configuration parsing, where I believe we will have to live with custom implementing both ways ... but happy to find a way here :).

Reply:
actually, I might be wrong and strum supports multiple string variations on parsing.

Reply:
it does :)

Reply:
Does it also have a way to iterate over all variants, so the unit test stays up to date automatically as new features land?

Reply:
Good call - several actually. I opted for exposing the string representation as `ReaderFeatures::VARIANTS` via `VaraintNames` and then include checks that we cover all variants in the roundtrip test. I thought this might be best, as we still have a way to see the variant names in code in case there are typos in the enum and tests also should be forced to stay in sync.